inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/PulsarSource.java

```diff
@@ -0,0 +1,312 @@
+    private boolean isRealTime = false;
+    private boolean isRestoreFromDB = false;
+
+    public PulsarSource() {
+    }
+
+    @Override
+    public void init(InstanceProfile profile) {
+        try {
+            LOGGER.info("PulsarSource init: {}", profile.toJsonStr());
+            this.profile = profile;
+            super.init(profile);
+            String cycleUnit = profile.get(TASK_CYCLE_UNIT);
+            if (cycleUnit.compareToIgnoreCase(CycleUnitType.REAL_TIME) == 0) {
+                isRealTime = true;
+                cycleUnit = CycleUnitType.HOUR;
+            }
+            queue = new LinkedBlockingQueue<>(CACHE_QUEUE_SIZE);
+            maxPackSize = profile.getInt(PROXY_PACKAGE_MAX_SIZE, DEFAULT_PROXY_PACKAGE_MAX_SIZE);
+            taskId = profile.getTaskId();
+            instanceId = profile.getInstanceId();
+            topic = profile.getInstanceId();
+            serviceUrl = profile.get(TASK_PULSAR_SERVICE_URL);
+            scanStartupMode = profile.get(TASK_PULSAR_SCAN_STARTUP_MODE);
+            timestamp = profile.getLong(TASK_PULSAR_RESET_TIME, 0);
+            pulsarClient = PulsarClient.builder().serviceUrl(serviceUrl).build();
+
+            EXECUTOR_SERVICE.execute(run());
+        } catch (Exception ex) {
+            stopRunning();
+            throw new FileException("error init stream for " + topic, ex);
+        }
+    }
+
+    private Runnable run() {
+        return () -> {
+            AgentThreadFactory.nameThread("pulsar-source-" + taskId + "-" + instanceId);
+            running = true;
+            try {
+                try (Consumer<byte[]> consumer = pulsarClient.newConsumer(Schema.BYTES)
+                        .topic(topic)
+                        .subscriptionName("inlong-agent-" + taskId)
+                        .subscriptionInitialPosition(SubscriptionInitialPosition.valueOf(scanStartupMode))
+                        .subscriptionType(SubscriptionType.Exclusive)
+                        .subscribe()) {
+
+                    if (!isRestoreFromDB && timestamp != 0L) {
+                        consumer.seek(timestamp);
+                        LOGGER.info("Reset consume from {}", timestamp);
+                    } else {
+                        LOGGER.info("Skip to reset consume");
+                    }
+
+                    doRun(consumer);
+                }
+            } catch (Throwable e) {
+                LOGGER.error("do run error maybe pulsar client is configured incorrectly: ", e);
+            }
+            running = false;
+        };
+    }
+
+    private void doRun(Consumer<byte[]> consumer) throws PulsarClientException {
+        long lastPrintTime = 0;
+        while (isRunnable()) {
+            boolean suc = waitForPermit(AGENT_GLOBAL_READER_SOURCE_PERMIT, BATCH_TOTAL_LEN);
+            if (!suc) {
+                break;
+            }
+            org.apache.pulsar.client.api.Message<byte[]> message = consumer.receive(0, TimeUnit.MILLISECONDS);
```

Comment:
If the timeout parameter is 0 and the topic has no messages backlog, `message` will be returned as null, then loop the next round immediately, this may cause high CPU load.

Reply:
If `message` returns null, `AgentUtils.silenceSleepInSeconds(1)` will let this thread sleep for 1 second and then loop next time.

Reply:
you're right.